### PR TITLE
Use a time-to-live cache for participation states to remove stale ritual participation state information

### PR DESCRIPTION
**Pull request by derekpierre**
**Type of PR:**
- [ ] Bugfix
- [ ] Feature
- [ ] Documentation
- [X] Other

**Required reviews:** 
- [ ] 1
- [X] 2
- [ ] 3

**What this does:**
Builds on #3183  (based over it).

One concern during a walkthrough with @KPrasch was that the `ParticipationState` dictionary is linearly increasing as rituals are initiated, and never cleared. This can be a very slow memory leak.

This PR uses a time-to-live cache utility for storing participation states and is periodically purged once entries are expired. ParticipationStates are cached for a little longer than the ritual timeout (an additional buffer (1hr in this PR)) just to be sure that we don't potentially get a repeated event right after purging. Remember that the EventScanner goes back some blocks when scanning in case there is a blockchain reorg, so we can get repeated events.

The overall goal here is to ensure that the participation states tracking mechanism used the `ActiveRitualTracker` periodically clears stale ritual state information.

**Issues fixed/closed:**
> - Fixes #...

**Why it's needed:**
> Explain how this PR fits in the greater context of the NuCypher Network.
> E.g., if this PR address a `nucypher/productdev` issue, let reviewers know!

**Notes for reviewers:**
> What should reviewers focus on? 
> Is there a particular commit/function/section of your PR that requires more attention from reviewers?

Take note of the timeout numbers, tough to guess them, but if you have some intuition of a better value, please speak up.

### CODE REVIEW

File: tests/unit/test_ttl_cache.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+
+
+def test_cache_get_remove_non_existent_entry():
+    ttl_cache = TTLCache(ttl=60)
+
+    # try to get non-existent entry
+    assert ttl_cache[1] is None
+
+    # try to remove non-existent entry
+    ttl_cache.remove(1)
```
Comment (manumonti) [inline tests/unit/test_ttl_cache.py:35]: Sorry, I don't get how this line works. Should be captured here an exception or something like that about we are trying to delete a non-existent entry?

**Review by manumonti — APPROVED** [review]
I've left some irrelevant comments, but overall LGTM! ✌️

File: nucypher/utilities/cache.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+                value = ttl_entry.value
+                if value:
+                    key_value_pairs.append((key, ttl_entry.value))
+                else:
+                    # expired entry, opportunity to remove it
+                    del self.__cache[key]
+
+        return key_value_pairs
+
+    def pop(self, key, default=None):
```
[the remark this answers is not in the document]
Comment (derekpierre) [reply nucypher/utilities/cache.py:94]: Stylistic really, that the caller can pass a different default if needed. Python `dict` objects provide the same method.

File: nucypher/blockchain/eth/trackers/dkg.py [hunk 3]
```diff
@@ -123,7 +128,17 @@ def __init__(
         )
 
         self.task = EventScannerTask(scanner=self.scan)
-        self._participation_states = dict()  # { ritual_id -> ParticipationState }
+
+        ritual_timeout = self.coordinator_agent.get_timeout()
+        # what's the buffer for potentially receiving repeated events - one hour?
```
[the remark this answers is not in the document]
Comment (derekpierre) [reply nucypher/blockchain/eth/trackers/dkg.py:133]: 👍 A logical calc like that is what I was hoping others would put forward.

Although I assume you mean 12s * 6 = 72 seconds (not minutes) = 1.2 minutes

I guess some of it depends on the EventScanner as well. It uses a `chain reorg scan window` value that specifies the number of blocks to go back before starting to scan again (hence the potential for repeated events). This value is currently 10 blocks - https://github.com/nucypher/nucypher/blob/development/nucypher/blockchain/eth/trackers/dkg.py#L23

Assuming 10 blocks * 12s = 120 seconds = 2 minutes.

So based on our discussion thus far, my use of 1 hour seems ridiculous 😅 ... maybe something more like 10-20 minutes is sufficient...?

File: nucypher/blockchain/eth/trackers/dkg.py [hunk 4]
```diff
@@ -68,6 +70,9 @@ class ActiveRitualTracker:
 
     MAX_CHUNK_SIZE = 10000
 
+    # how often to check/purge for expired cached values - 8hrs?
```
[the remark this answers is not in the document]
Comment (derekpierre) [reply nucypher/blockchain/eth/trackers/dkg.py:73]: 👍 The tricky is part is are we happy with this way of purging these items. In this case we're piggy-backing on ritual events instead of having a dedicated task. Using a dedicated periodic running task just felt excessive.

The one edge case for this current logic is if a bunch of rituals get created initially (like hundreds etc.) before the 8-hour timeline, and then no more rituals afterwards. Without subsequent events the cache doesn't clear, unless of course the node restarts. Not the end of the world in my mind, it's 3 booleans stored per ritual which is pretty small.

File: nucypher/blockchain/eth/trackers/dkg.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (manumonti) [reply nucypher/blockchain/eth/trackers/dkg.py]: > Although I assume you mean 12s * 6 = 72 seconds (not minutes) = 1.2 minutes

Wops, you're right.

> So based on our discussion thus far, my use of 1 hour seems ridiculous 😅 ... maybe something more like 10-20 minutes is sufficient...?

I am more inclined to 10 minutes than 20 minutes. 10 minutes window is equivalent to 50 blocks scan window, which sounds enough to me.

File: nucypher/blockchain/eth/trackers/dkg.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (manumonti) [reply nucypher/blockchain/eth/trackers/dkg.py:73]: Good reasoning. 8 hrs looks fine for this edge case. And as you say, the size of the stored data is pretty small, I don't think we will have problems even with larger periods of time.

File: tests/unit/test_ttl_cache.py (hunk 1 above)
Comment (derekpierre) [reply tests/unit/test_ttl_cache.py:35]: Fair question. 

It's really ensuring that a `KeyError` isn't raised.

Remember that cache entries can be removed (in the backgrounds) if they are expired, so for `remove()`, one of the following scenarios is possible:
a) Ask to remove a keyed entry that is currently in the cache
b) Ask to remove a keyed entry that was never in the cache
c) Ask to remove a keyed entry that is in the cache but expired
d) Ask to remove a keyed entry that was in the cache but was auto-removed previously because it was expired (eg. someone called length on the cache or something that performed a purge)

- a) is the regular case - so remove the item, great
- for b) we could throw a `KeyError` but how do we tell the difference between b) and scenario d)
- for c) would you do nothing or `KeyError`? If `KeyError` what's the difference between scenario c) and d) - just the timing of when you happen to call `remove()`? That's pretty inconsistent.

So to be easily consistent, we just never return a `KeyError` and instead do nothing if someone tries to remove a non-existent entry. That philosophy holds for the cache in general i.e. no `KeyError` is raised - same for the `__getitem__()` functionality.

File: nucypher/blockchain/eth/trackers/dkg.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (derekpierre) [reply nucypher/blockchain/eth/trackers/dkg.py]: ✔️  - 10 minutes it is!

File: nucypher/blockchain/eth/trackers/dkg.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/blockchain/eth/trackers/dkg.py:73]: Considering this cache only grows in size with new DKG activities, piggy-backing is still an effective state pruning strategy here 👍🏻

File: nucypher/blockchain/eth/trackers/dkg.py [hunk 5]
```diff
@@ -123,7 +131,17 @@ def __init__(
         )
 
         self.task = EventScannerTask(scanner=self.scan)
-        self._participation_states = dict()  # { ritual_id -> ParticipationState }
+
+        cache_ttl = (
+            self.coordinator_agent.get_timeout()
+            + self._RITUAL_TIMEOUT_ADDITIONAL_TTL_BUFFER
+        )
+        self._participation_states = TTLCache(
```
Comment (KPrasch) [inline nucypher/blockchain/eth/trackers/dkg.py:139]: Ah yes, return of the `TTLCache` :-)

**Review by KPrasch — APPROVED** [review]
🤠

File: nucypher/blockchain/eth/trackers/dkg.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (derekpierre) [reply nucypher/blockchain/eth/trackers/dkg.py:73]: 👍 - great point.

File: nucypher/blockchain/eth/trackers/dkg.py (hunk 5 above)
Comment (derekpierre) [reply nucypher/blockchain/eth/trackers/dkg.py:139]: Haha yes - you remembered! I indeed revived it from the old Porter WIP PR, and made some improvements.

File: tests/unit/test_ttl_cache.py (hunk 1 above)
Comment (manumonti) [reply tests/unit/test_ttl_cache.py:35]: Got it, thanks for your nice clarification
